tensorstore/kvstore/s3/s3_key_value_store.cc

```diff
@@ -0,0 +1,1306 @@
+      execution::set_stopping(receiver_);
+    }
+    return absl::OkStatus();
+  }
+};
+
+void S3KeyValueStore::ListImpl(ListOptions options,
+                                AnyFlowReceiver<absl::Status, Key> receiver) {
+  s3_list.Increment();
+  if (options.range.empty()) {
+    execution::set_starting(receiver, [] {});
+    execution::set_done(receiver);
+    execution::set_stopping(receiver);
+    return;
+  }
+
+  auto state = internal::MakeIntrusivePtr<ListTask>(
+      IntrusivePtr<S3KeyValueStore>(this), std::move(options),
+      std::move(receiver),
+      /*resource=*/tensorstore::StrCat(endpoint_, "/"));
+
+  intrusive_ptr_increment(state.get());  // adopted by ListTask::Start.
+  read_rate_limiter().Admit(state.get(), &ListTask::Start);
+}
+
+// Receiver used by `DeleteRange` for processing the results from `List`.
+struct DeleteRangeListReceiver {
+  IntrusivePtr<S3KeyValueStore> owner_;
+  Promise<void> promise_;
+  FutureCallbackRegistration cancel_registration_;
+
+  void set_starting(AnyCancelReceiver cancel) {
+    cancel_registration_ = promise_.ExecuteWhenNotNeeded(std::move(cancel));
+  }
+
+  void set_value(std::string key) {
+    assert(!key.empty());
+    if (!key.empty()) {
+      LinkError(promise_, owner_->Delete(std::move(key)));
+    }
+  }
+
+  void set_error(absl::Status error) {
+    SetDeferredResult(promise_, std::move(error));
+    promise_ = Promise<void>();
+  }
+
+  void set_done() { promise_ = Promise<void>(); }
+
+  void set_stopping() { cancel_registration_.Unregister(); }
+};
+
+Future<const void> S3KeyValueStore::DeleteRange(KeyRange range) {
+  s3_delete_range.Increment();
+  if (range.empty()) return absl::OkStatus();
+
+  // TODO(jbms): It could make sense to rate limit the list operation, so that
+  // we don't get way ahead of the delete operations.  Currently our
+  // sender/receiver abstraction does not support back pressure, though.
+  auto op = PromiseFuturePair<void>::Make(tensorstore::MakeResult());
+  ListOptions list_options;
+  list_options.range = std::move(range);
+  ListImpl(list_options, DeleteRangeListReceiver{
+                             internal::IntrusivePtr<S3KeyValueStore>(this),
+                             std::move(op.promise)});
+  return std::move(op.future);
+}
+
+Result<kvstore::Spec> ParseS3Url(std::string_view url) {
+  auto parsed = internal::ParseGenericUri(url);
+  assert(parsed.scheme == kUriScheme);
+  if (!parsed.query.empty()) {
+    return absl::InvalidArgumentError("Query string not supported");
+  }
+  if (!parsed.fragment.empty()) {
+    return absl::InvalidArgumentError("Fragment identifier not supported");
+  }
+  size_t end_of_bucket = parsed.authority_and_path.find('/');
+  std::string_view bucket = parsed.authority_and_path.substr(0, end_of_bucket);
+  if (!IsValidBucketName(bucket)) {
+    return absl::InvalidArgumentError(
+        tensorstore::StrCat("Invalid S3 bucket name: ", QuoteString(bucket)));
+  }
+  std::string_view path =
```

Comment:
path needs percent decoding

Reply:
Encoding I believe? I changed this to use `S3UriObjectKeyEncode`. Also, as I write this, is this conversion necessary at this point (as the key will be encoded in the various Tasks)?

Reply:
Here we are parsing the tensorstore kvstore "s3://" URL syntax, which is independent of the URL syntax for actual s3 requests.

`ParseGenericUrl` separates the components of the URL but does not handle percent decoding.  For example, if the bucket is "my-bucket" and the path is "my path", it will be encoded as "s3://my-bucket/my%20path".  But in the kvstore Spec, the path should be represented as just "my path".  Therefore we need percent decoding of `path`, as is done in the GCS driver.  When later the request is made to S3, it will indeed be re-encoded as "my%20path".  But if we didn't percent decode here, then it would instead be "double percent encoded" as "my%2520path" and refer to a different object.

Reply:
Thanks, used percent decoding.